Fixed most variableScope hints by CppCheck 1.86

SunBlack opened the pull request:
There are 3 warnings left, which I didn't fixed, because it currently has a better readability.

Addiotnally: There are a lot of variableScope fixes, which are not detected by CppCheck. Seems CppCheck ignore hpp files and loops :(. Adjustments to loops are necessary, because Clang-tidy ignore them during modernize-loop-convert, if iterators are not defined within loop head.

Skipped doc/tutirals und OpenNurbs directory.

PS: Sorry for the big changes hog.cpp. I formated code to understand what CppCheck see there and after this I simplified duplicated code

SunBlack commented on apps/in_hand_scanner/src/icp.cpp:
```diff
@@ -404,8 +404,7 @@ pcl::ihs::ICP::selectDataPoints (const CloudXYZRGBNormalConstPtr& cloud_data) co
 
-  CloudXYZRGBNormal::const_iterator it_in = cloud_data->begin ();
-  for (; it_in!=cloud_data->end (); ++it_in)
+  for (auto it_in = cloud_data->begin (); it_in!=cloud_data->end (); ++it_in)
```
```
apps/in_hand_scanner/src/icp.cpp:407:33: error: no member named 'cbegin' in 'pcl::PointCloud<pcl::PointXYZRGBNormal>'; did you mean 'begin'?
  for (auto it_in = cloud_data->cbegin (); it_in!=cloud_data->cend (); ++it_in)
```

And pointcloud.h:
```
      inline iterator begin () { return (points.begin ()); }
      inline iterator end ()   { return (points.end ()); }
      inline const_iterator begin () const { return (points.begin ()); }
      inline const_iterator end () const  { return (points.end ()); }
```
`cloud_data` is const, so he should use correct version.

SergioRAgostinho replied:
I think I'll open an issue to rename these methods to `cbegin` and `cend`.

SunBlack commented on apps/in_hand_scanner/src/input_data_processing.cpp:
```diff
@@ -244,7 +243,7 @@ pcl::ihs::InputDataProcessing::calculateNormals (const CloudXYZRGBAConstPtr& clo
   invalid_pt.data_n [3] = 0.f;
 
-  for (; it_in!=cloud_in->end (); ++it_in, ++it_n, ++it_out)
+  for (auto it_in = cloud_in->begin (); it_in!=cloud_in->end (); ++it_in, ++it_n, ++it_out)
```
Same like above

SunBlack commented on apps/point_cloud_editor/src/cloud.cpp:
```diff
@@ -193,9 +193,8 @@ Cloud::setSelection (SelectionPtr selection_ptr)
-  Selection::const_iterator it;
   // assumes selection is sorted small to large
-  for (it = selection_ptr->begin(); it != selection_ptr->end(); ++it, ++pos)
+  for (auto it = selection_ptr->begin(); it != selection_ptr->end(); ++it, ++pos)
```
`Selection::begin` returns always a `const_iterator`

SunBlack commented on apps/point_cloud_editor/src/cloud.cpp:
```diff
@@ -330,8 +329,7 @@ void
   unsigned int pos = cloud_.size();
-  Selection::const_reverse_iterator rit;
-  for (rit = selection.rbegin(); rit != selection.rend(); ++rit)
+  for (auto rit = selection.rbegin(); rit != selection.rend(); ++rit)
```
`Selection::begin` returns always a `const_iterator`

SunBlack commented on apps/point_cloud_editor/src/cloud.cpp:
```diff
@@ -442,8 +440,7 @@ Cloud::restore (const CopyBuffer& copy_buffer, const Selection& selection)
   unsigned int pos = cloud_.size();
-  Selection::const_reverse_iterator rit;
-  for (rit = selection.rbegin(); rit != selection.rend(); ++rit)
+  for (auto rit = selection.rbegin(); rit != selection.rend(); ++rit)
```
`Selection::rbegin` returns always a `const_reverse_iterator`

SunBlack commented on apps/point_cloud_editor/src/copyBuffer.cpp:
```diff
@@ -60,8 +60,7 @@ CopyBuffer::set (ConstCloudPtr cloud_ptr, const Selection& selection)
       return;
-  Selection::const_iterator s_it;
-  for(s_it = selection.begin(); s_it != selection.end(); ++s_it)
+  for(auto s_it = selection.begin(); s_it != selection.end(); ++s_it)
```
`Selection::begin` returns always a `const_iterator`

SunBlack commented on apps/point_cloud_editor/src/selection.cpp:
```diff
@@ -72,8 +72,7 @@ Selection::addIndex (const IndexVector &indices)
 {
-  IndexVector::const_iterator it;
-  for(it = indices.begin(); it != indices.end(); ++it)
+  for(auto it = indices.begin(); it != indices.end(); ++it)
```
`IndexVector` is `const`, so already `const_iterator` is used. Nevertheless I changed it.

SunBlack commented on apps/point_cloud_editor/src/transformCommand.cpp:
```diff
@@ -131,14 +129,13 @@ TransformCommand::undo()
   // now modify the selected points' coordinates
-  Selection::const_iterator it;
-  for(it = sel_ptr -> begin(); it != sel_ptr-> end(); ++it)
+  for(auto it = sel_ptr -> begin(); it != sel_ptr-> end(); ++it)
```
`Selection::begin` returns always a `const_iterator`

SergioRAgostinho commented:
> @SergioRAgostinho thanks for the review! Can I help somehow (without duplicating the effort)?

Right now, you can double check the comments in which I claim "that functionality is being changed". I think this is one or two points. If you confirm that nothing is changed then mark the comment as resolved and it's less work for everyone. Some fresh eyes into these might help clear any mistakes I might have made.

SunBlack commented on filters/include/pcl/filters/impl/crop_hull.hpp:
```diff
@@ -136,12 +135,12 @@ pcl::CropHull<PointT>::applyFilter2D (PointCloud &output)
         // once a point has tested +ve for being inside one polygon, we can
         // stop checking the others:
-        break;
+        return;
```
No, but you are right. That's an error. Just saw `poly == hull_polygons_.size ()` will be only true, in case the `break` was never called. And in case `break` is called, nothing further will be done. Didn't saw the outer loop, so `return` isn't correct.

SunBlack commented on apps/point_cloud_editor/src/transformCommand.cpp:
```diff
@@ -76,18 +76,16 @@ TransformCommand::undo()
-  unsigned int index = 0;
-  Selection::const_iterator it;
-  for(it = internal_selection_ptr_ -> begin();
+  for(auto it = internal_selection_ptr_ -> begin();
```
`Selection::begin` returns always a `const_iterator`

SunBlack commented on registration/include/pcl/registration/impl/gicp.hpp:
```diff
@@ -64,9 +64,8 @@ pcl::GeneralizedIterativeClosestPoint<PointSource, PointTarget>::computeCovarian
-  typename pcl::PointCloud<PointT>::const_iterator points_iterator = cloud->begin ();
   MatricesVector::iterator matrices_iterator = cloud_covariances.begin ();
-  for(;
+  for(auto points_iterator = cloud->begin ();
```
`PointCloud` has `cbegin`. But `cloud` is const, so he should use correct version.

taketwo commented on recognition/src/ransac_based/orr_octree_zprojection.cpp:
```diff
@@ -98,24 +99,24 @@ pcl::recognition::ORROctreeZProjection::build (const ORROctree& input, float eps
+  full_leaves_bounds[0] = std::numeric_limits<float>::max();
+  full_leaves_bounds[1] = std::numeric_limits<float>::lowest();
+  full_leaves_bounds[2] = std::numeric_limits<float>::max();
+  full_leaves_bounds[3] = std::numeric_limits<float>::lowest();
```
Agree. Also note that in the original code the for loop starts with `++fl_it`, meaning that in the first iteration the iterator is pointing to `begin + 1`, not `begin` as in your modified code.

SunBlack replied:
Moved to #2812

SunBlack commented on geometry/include/pcl/geometry/mesh_base.h:
```diff
@@ -287,13 +287,10 @@ namespace pcl
           {
-            typename EdgeDataCloud::const_iterator it_ed_old = edge_data_cloud_.begin ();
-            typename EdgeDataCloud::iterator       it_ed_new = edge_data_cloud_.begin ();
+            auto it_ed_old = edge_data_cloud_.begin ();
```
`PointCloud` has no `cebgin`

SunBlack commented on ml/src/svm.cpp:
```diff
@@ -2124,9 +2104,6 @@ static void sigmoid_train (
     }
   }
 
-  if (iter >= max_iter)
```
True, I just read `iter > max_iter` instead of `iter >= max_iter`

SunBlack commented:
@taketwo: This PR needs already a merge, so I believe it is better to merge #2812 first, then merge and if all is reviewed maybe squash to keep log clean of reverted changes.

SergioRAgostinho commented:
For my side, this is ready for a rebase. Please wait for @taketwo 's green light just in case.

SunBlack commented:
So I can merge and rebase/squash now?

taketwo commented:
Yes! Please also fix the conflict in "visualization/src/pcl_visualizer.cpp", otherwise we can not merge.

SergioRAgostinho commented:
I merged #2812. Let's go for the final rebase/squash. :+1:

SunBlack commented:
Needed to integrate some fixes
* I had to revert changes to `ply_parser` (fixes test issues)
* Adjusted `doc/tutorials/.../supervoxel_clustering` used invalid `cbegin` for `adjacent_supervoxel_centers` => `begin`
* Adjusted lines in `supervoxel_clustering.rst`

taketwo commented:
@SergioRAgostinho merging?

SergioRAgostinho commented:
I didn't check the very last changes, but now they're squashed anyway.  The test were failing and now they're not so everything is fine by me ^^

taketwo commented:
> Next time, please split these over multiple PRs, focusing on a couple of modules in each. Take the threshold of 500 touched lines of code as a good rule of thumb to decide when to split.

@SunBlack thanks for the effort with this (and all the other PRs). However, please do keep this request in mind.
